test/test_autograd.py

```diff
@@ -3568,6 +3575,54 @@ def backward(ctx, input):
         # Verify callback is called only once.
         self.assertEqual(1, counter[0])
 
+    def test_reentrant_with_leaf_variable_hook(self):
+        handle = None
+        param = torch.rand(10, requires_grad=True)
+
+        def add_gradient_penalty_to_grad(grad):
+            handle.remove()
+            old_param_grad = grad
+            param.grad = None
+            # Add some sort of gradient penalty by directly updating the gradients
+            with torch.enable_grad():
+                g = grad.detach().requires_grad_()
+                new_param = param.detach().requires_grad_()
```

Comment:
This is needed, otherwise we will ended up with `param`'s AccumulateGrad that is sharing between two GraphTask(one in parent, one in hook), so when you run the AccumulateGrad hook for leaf variable `param`, we are using `param` as the leaf variable for backward call in the hook as well. That will lead to the backward in the hook also tries to run the same `param`'s AccumulateGrad node, which will break thread safety combined with the new Autograd PR.

Reply:
But that should work no?
In particular for things that were discussed in the distributed meeting that were considering moving optimizers into hooks, then trying to add any penalty to the original weights will just deadlock?

Reply:
So after offline discussions, this is fine.